app/src/main/java/at/roteskreuz/stopcorona/model/repositories/ReportingRepository.kt

```diff
@@ -132,22 +141,95 @@ class ReportingRepositoryImpl(
-        temporaryExposureKeys: List<TemporaryExposureKey>): MessageType.InfectionLevel {
+    private suspend fun uploadInfectionInfo(): MessageType.InfectionLevel {
         return withContext(coroutineContext) {
-            val infectionLevel = messageTypeSubject.value as? MessageType.InfectionLevel ?: throw InvalidConfigurationException.InfectionLevelNotSet
+            val infectionLevel = messageTypeSubject.value as? MessageType.InfectionLevel
+                ?: throw InvalidConfigurationException.InfectionLevelNotSet
+
+            val configuration = configurationRepository.observeConfiguration().blockingFirst()
+            val uploadKeysDays = configuration.uploadKeysDays
+                ?: throw InvalidConfigurationException.NullNumberOfDaysToUpload
+            var thresholdTime = ZonedDateTime.now()
+                .minusDays(uploadKeysDays.toLong())
+                .startOfTheDay()
+                .toRollingStartIntervalNumber()
+
+            val infectionMessages = mutableListOf<TemporaryExposureKeysWrapper>()
+
+            if (infectionLevel == MessageType.InfectionLevel.Red) {
+                val sentTemporaryExposureKeys =
+                    infectionMessengerRepository.getSentTemporaryExposureKeysByMessageType(
+                        MessageType.InfectionLevel.Yellow
+                    )
+
+                infectionMessages.addAll(
+                    sentTemporaryExposureKeys
+                        .map { message ->
+                            TemporaryExposureKeysWrapper(
+                                message.rollingStartIntervalNumber,
+                                message.password,
+                                MessageType.InfectionLevel.Red
+                            )
+                        }
+                )
+
+                if (infectionMessages.isNotEmpty()) {
+                    infectionMessages.sortByDescending { content -> content.rollingStartIntervalNumber }
+                    thresholdTime = infectionMessages.first().rollingStartIntervalNumber
+                }
+            } else if (infectionLevel == MessageType.InfectionLevel.Yellow) {
+                val resetMessages =
+                    infectionMessengerRepository.getSentTemporaryExposureKeysByMessageType(
+                        MessageType.InfectionLevel.Yellow
+                    )
+                        .map { temporaryExposureKey ->
+                            TemporaryExposureKeysWrapper(
+                                temporaryExposureKey.rollingStartIntervalNumber,
+                                temporaryExposureKey.password,
+                                MessageType.Revoke.Suspicion
+                            )
+                        }
+
+                infectionMessengerRepository.storeSentTemporaryExposureKeys(resetMessages)
+            }
+
+            val temporaryExposureKeysFromSDK =
+                exposureNotificationRepository.getTemporaryExposureKeys()
```

Comment:
I guess it might give us an error if the framework is not running, correct @falkorichter ?
We should check before
```
                if (exposureNotificationManager.currentPhase.let { it is FrameworkRunning || it is NotCritical }.not()) {
                    throw error // framework is not ready
                }
```

Please apply for all use cases where getTemporaryExposureKeys is used.

Reply:
@dusanjencik , @falkorichter 
Do we need it at this place? 
Wouldn't trigger an ApiException? such exceptions are handled in `uploadData` method in `ReportingStatusViewModel`

Reply:
Let's discuss it together.